drivers/random/random_stm32_rng.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2017 Erwin Rol <erwin@erwinrol.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <random.h>
+#include <drivers/rand32.h>
+#include <init.h>
+#include <misc/__assert.h>
+#include <misc/util.h>
+#include <errno.h>
+#include <soc.h>
+#include <misc/printk.h>
+#include <clock_control.h>
+#include <clock_control/stm32_clock_control.h>
+
+struct random_stm32_rng_dev_cfg {
+	struct stm32_pclken pclken;
+};
+
+struct random_stm32_rng_dev_data {
+	RNG_HandleTypeDef hrng;
+	struct device *clock;
+};
+
+#define DEV_DATA(dev) \
+	((struct random_stm32_rng_dev_data *)(dev)->driver_data)
+
+#define DEV_CFG(dev) \
+	((struct random_stm32_rng_dev_cfg *)(dev)->config->config_info)
+
+
+static int random_stm32_rng_get_entropy(struct device *dev, u8_t *buffer,
+					u16_t length)
+{
+	struct random_stm32_rng_dev_data *dev_data;
+
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG(buffer != NULL);
+
+	dev_data = DEV_DATA(dev);
+
+	__ASSERT_NO_MSG(dev_data != NULL);
+
+	int n = sizeof(u32_t);
+
+	while (length > 0) {
+		u32_t rndbits;
+		u8_t *p_rndbits = (u8_t *)&rndbits;
+
+		if (HAL_RNG_GenerateRandomNumber(&(dev_data->hrng), &rndbits) != HAL_OK) {
+			return -EIO;
+		}
+
+		if (length < sizeof(u32_t))
+			n = length;
+
+		for (int i = 0; i < n; i++) {
+			*buffer++ = *p_rndbits++;
```

Comment:
Ah, I think I was thinking that the types of buffer & p_rndbits where u32_t, seeing they are u8_t.